include/seqan3/search/detail/unidirectional_search_algorithm.hpp

```diff
@@ -37,17 +41,144 @@ enum class error_type : uint8_t
+ * \tparam index_t The type of index; must model seqan3::fm_index_specialisation.
+ */
+template <typename configuration_t, fm_index_specialisation index_t, typename ...policies_t>
+class unidirectional_search_algorithm : public policies_t...
+{
+private:
+    //!\brief The search configuration traits.
+    using traits_t = search_traits<configuration_t>;
+
+public:
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    unidirectional_search_algorithm() = default; //!< Defaulted.
+    unidirectional_search_algorithm(unidirectional_search_algorithm const &) = default; //!< Defaulted.
+    unidirectional_search_algorithm(unidirectional_search_algorithm &&) = default; //!< Defaulted.
+    unidirectional_search_algorithm & operator=(unidirectional_search_algorithm const &) = default; //!< Defaulted.
+    unidirectional_search_algorithm & operator=(unidirectional_search_algorithm &&) = default; //!< Defaulted.
+    ~unidirectional_search_algorithm() = default; //!< Defaulted.
+
+    /*!\brief Constructs from a configuration object and an index.
+     * \tparam configuration_t The search configuration type.
+     * \tparam index_t The type of index; must model seqan3::fm_index_specialisation.
+     * \param[in] cfg The configuration object that guides the search algorithm.
+     * \param[in] index The index used in the algorithm.
+     */
+    unidirectional_search_algorithm(configuration_t const & cfg, index_t const & index) : policies_t{}...
+    {
+        config = cfg;
+        index_ptr = &index;
+    }
+    //!\}
+
+    /*!\brief Searches a query sequence in an FM index using trivial backtracking.
+     * \tparam query_t       Must model std::ranges::input_range over the index's alphabet.
+     * \param[in] query      Query sequence to be searched in the index.
+     *
+     * ### Complexity
+     *
+     * \f$O(|query|^e)\f$ where \f$e\f$ is the maximum number of errors.
+     */
+    template <typename query_t>
+    auto operator()(query_t && query) noexcept
+    {
+        auto error_state = this->max_error_counts(config, query); // see policy_max_error
+
+        // construct internal delegate for collecting hits for later filtering (if necessary)
+        std::vector<typename index_t::cursor_type> internal_hits;
+        delegate = [&internal_hits] (auto const & it)
+        {
+            internal_hits.push_back(it);
+        };
+
+        perform_search_by_mode(internal_hits, query, error_state);
+
+        return this->make_results(std::move(internal_hits), config); // see policy_result_builder
+    }
+
+private:
+    //!\brief The configuration object.
+    configuration_t config{};
+
+    //!\brief A pointer to the fm index which is used to perform the unidirectional search.
+    index_t const * index_ptr{nullptr};
+
+    //!\brief A function object that stores the on-hit-delegate to be executed whenever a hit in the index is found.
+    std::function<void(typename index_t::cursor_type const &)> delegate;
```

Comment:
just out of curiosity, why is this needed? In the other implementation you pass it into the perform_search function, which seems plausible to do.

Reply:
I was trying to reduce the parameter list of the free functions. search_trivial is fine but when you look into the free functions in `search_scheme_algorithm.hpp` (the ones I haven't touched yet) the number of parameters is huge. So this was rather a trial here, as I thought it is equivalent, but might serve as an example for further refactoring in  `search_scheme_algorithm.hpp`. If it is not equivalent but the function parameter is preferable I have no strong feelings to change it back.